File: lib/shortbread/extraction.rb
# frozen_string_literal: true

module Shortbread
  # Visible text of a document, plus the map back to the bytes it came from. Anchors target this
  # text rather than source bytes, so markup changes that do not change what a reader sees leave
  # the Anchor undisturbed.
  #
  # Visibility is decided by TAG NAME ONLY. CSS is never consulted, so `display:none` text is
  # extracted. That is deliberate: the TypeScript twin walks the DOM under the same rule, and it
  # is agreement between the two that keeps client and server offsets in lockstep. An extractor
  # that consulted CSS — `innerText`, `getClientRects` — would silently break server-side
  # verification for every Anchor after a hidden element.
  module Extraction
    Extracted = Data.define(:text, :map)

    INVISIBLE = %w[script style head title meta link].freeze

    BLOCK = %w[
      p div h1 h2 h3 h4 h5 h6 li tr br hr section article header footer pre blockquote
      table thead tbody td th ul ol
    ].freeze

    ENTITIES = {
      "&amp;" => "&", "&lt;" => "<", "&gt;" => ">", "&quot;" => '"',
      "&#39;" => "'", "&apos;" => "'", "&nbsp;" => " ", "&mdash;" => "—", "&hellip;" => "…"
    }.freeze

    MAX_ENTITY_LENGTH = 10

    module_function

    # The visible text of one Manifest Entry, read through the Blob store's verification so the
    # bytes an Anchor is checked against are provably the bytes the Release was published with.
    # Returns nil when the Blob cannot be read, leaving the caller to decide what that means.
    def for_entry(entry, blob_store:)
      io = blob_store.open_verified(
        storage_key: entry.blob.storage_key,
        sha256: entry.blob.sha256,
        byte_size: entry.byte_size
      )
      from_html(io.read.force_encoding(Encoding::UTF_8))
    rescue BlobStore::ContentMismatch, BlobStore::StorageFailure
      nil
    ensure
      io&.close
    end

    # Where in the original bytes a text offset came from, so an Anchor resolved in text space can
    # be pointed back at the content-addressed Blob.
    def source_offset(extracted, text_offset)
      extracted.map[text_offset]
    end

    def from_html(html)
      state = { text: +"", map: [], pre_depth: 0, skip_until: nil }
      index = 0

      while index < html.length
        if html[index] == "<"
          close = html.index(">", index)
          break unless close

          apply_tag(state, html[(index + 1)...close].to_s, index)
          index = close + 1
          next
        end

        index = state[:skip_until] ? index + 1 : append_character(state, html, index)
      end

      trimmed(state)
    end

    def apply_tag(state, tag, index)
      closing = tag.start_with?("/")
      name = tag[%r{\A/?\s*([a-zA-Z0-9]+)}, 1]&.downcase

      if state[:skip_until]
        state[:skip_until] = nil if closing && name == state[:skip_until]
        return
      end

      if INVISIBLE.include?(name) && !closing
        state[:skip_until] = name unless tag.end_with?("/")
        return
      end

      if name == "pre"
        state[:pre_depth] += closing ? -1 : 1
        state[:pre_depth] = 0 if state[:pre_depth].negative?
      end

      # Both opening AND closing block tags break. Breaking only on entering an element lets the
      # whitespace after `</h1>` survive as a space and desyncs every subsequent offset.
      break_block(state, index) if BLOCK.include?(name)
    end

    def break_block(state, index)
      return if state[:text].empty? || state[:text].end_with?("\n")

      # A block break absorbs the collapsed space before it. Whitespace between a word and the end
      # of its block is not visible to a reader, so leaving it would make a reformatted document
      # extract differently from a minified one and shift every offset after the break.
      if state[:text].end_with?(" ")
        state[:text].chop!
        state[:map].pop
      end

      state[:text] << "\n"
      state[:map] << index
    end

    def append_character(state, html, index)
      character = html[index]

      if character == "&"
        semicolon = html.index(";", index)
        if semicolon && semicolon - index <= MAX_ENTITY_LENGTH
          decoded = ENTITIES[html[index..semicolon]]
          if decoded
            decoded.each_char { state[:map] << index }
            state[:text] << decoded
            return semicolon + 1
          end
        end
      end

      if character.match?(/\s/)
        append_whitespace(state, character, index)
        return index + 1
      end

      state[:text] << character
      state[:map] << index
      index + 1
    end

    # Runs of whitespace collapse to a single space the way a browser renders them — except
    # inside <pre>, where whitespace is content.
    def append_whitespace(state, character, index)
      if state[:pre_depth].positive?
        state[:text] << character
        state[:map] << index
      elsif !(state[:text].empty? || state[:text].end_with?(" ", "\n"))
        state[:text] << " "
        state[:map] << index
      end
    end

    # Trim text and map together — stripping only the text desyncs every offset after the
    # leading whitespace, which looks exactly like anchor drift.
    def trimmed(state)
      text = state[:text]
      leading = text[/\A\s*/].length
      trailing = text[/\s*\z/].length
      keep = text.length - leading - trailing
      keep = 0 if keep.negative?

      Extracted.new(text: text[leading, keep] || "", map: state[:map][leading, keep] || [])
    end
  end
end

File: app/frontend/anchoring/extraction.ts
/**
 * The TypeScript twin of `Shortbread::Extraction`, walking the DOM instead of scanning bytes.
 *
 * Visibility is decided by TAG NAME ONLY. `innerText`, `getClientRects`, and every other
 * CSS-aware test are deliberately avoided: they would drop `display:none` text that the Ruby
 * extractor keeps, and a measured 19-character divergence on one hidden block silently breaks
 * server-side verification for every Anchor after it. Emitting hidden text is harmless because
 * both sides do it identically — it is the agreement that matters, not the visibility.
 *
 * Offsets are Unicode code points, matching Ruby. A DOM node's `textContent` is UTF-16, so every
 * boundary crossing converts.
 */

export type Extracted = {
  text: string
  /** For each extracted code point, the DOM node and UTF-16 offset within it that produced it. */
  map: Array<{ node: Text; offset: number }>
}

const INVISIBLE = new Set(['script', 'style', 'head', 'title', 'meta', 'link'])

const BLOCK = new Set([
  'p', 'div', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'li', 'tr', 'br', 'hr', 'section', 'article',
  'header', 'footer', 'pre', 'blockquote', 'table', 'thead', 'tbody', 'td', 'th', 'ul', 'ol',
])

type State = {
  text: string[]
  map: Array<{ node: Text; offset: number }>
  preDepth: number
}

export const fromDom = (root: Node): Extracted => {
  const state: State = { text: [], map: [], preDepth: 0 }
  walk(root, state)
  return trimmed(state)
}

const walk = (node: Node, state: State): void => {
  if (node.nodeType === Node.TEXT_NODE) {
    appendText(node as Text, state)
    return
  }

  if (node.nodeType !== Node.ELEMENT_NODE) return

  const element = node as Element
  const name = element.tagName.toLowerCase()
  if (INVISIBLE.has(name)) return

  const isPre = name === 'pre'
  const isBlock = BLOCK.has(name)

  // Both ENTERING and LEAVING a block break the text. Breaking only on entering lets the
  // whitespace after `</h1>` survive as a space and desyncs every subsequent offset — the exact
  // divergence the browser spike found on its first run.
  if (isBlock) breakBlock(state)
  if (isPre) state.preDepth += 1

  for (const child of Array.from(element.childNodes)) walk(child, state)

  if (isPre) state.preDepth = Math.max(0, state.preDepth - 1)
  if (isBlock) breakBlock(state)
}

const breakBlock = (state: State): void => {
  if (state.text.length === 0) return
  if (state.text[state.text.length - 1] === '\n') return

  state.text.push('\n')
  // A synthesised break corresponds to a tag rather than to any selectable character. It is
  // anchored to the last real position so a selection landing here still maps somewhere sane.
  state.map.push(state.map[state.map.length - 1])
}

const appendText = (node: Text, state: State): void => {
  const content = node.data
  let utf16Offset = 0

  for (const character of content) {
    if (/\s/.test(character)) {
      appendWhitespace(character, node, utf16Offset, state)
    } else {
      state.text.push(character)
      state.map.push({ node, offset: utf16Offset })
    }
    utf16Offset += character.length
  }
}

/**
 * Runs of whitespace collapse to a single space the way a browser renders them — except inside
 * `<pre>`, where whitespace is content.
 */
const appendWhitespace = (
  character: string,
  node: Text,
  utf16Offset: number,
  state: State,
): void => {
  if (state.preDepth > 0) {
    state.text.push(character)
    state.map.push({ node, offset: utf16Offset })
    return
  }

  const last = state.text[state.text.length - 1]
  if (state.text.length === 0 || last === ' ' || last === '\n') return

  state.text.push(' ')
  state.map.push({ node, offset: utf16Offset })
}

/**
 * Trim text and map together — trimming only the text desyncs every offset after the leading
 * whitespace, which looks exactly like anchor drift.
 */
const trimmed = (state: State): Extracted => {
  let start = 0
  let end = state.text.length
  while (start < end && /\s/.test(state.text[start])) start += 1
  while (end > start && /\s/.test(state.text[end - 1])) end -= 1

  return {
    text: state.text.slice(start, end).join(''),
    map: state.map.slice(start, end),
  }
}
